fixes issue where external variables are not properly initialized

DanielYankura opened the pull request:
-issue was caused by external variables never being initialized 
-in InitProblemAction.C it now checks if there is an external variable 
-if there is then it calls initializeVariable
-initializeVariable was added to BasicExternalProblem and SyncTestExternalProblem

closes #18396

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
If a variable is not properly initialized then it can cause a lot of issues.
## Design
<!--A concise description (design) of the enhancement.-->
Adds check to InitProblemAction that looks to see if there are any external variables. If there are then it calls the initialzeVariable function which must be defined in the desired external variable source file.
## Impact
<!--Will the enhancement change existing APIs or add something new?-->
Allows users to utilize external variables for a broader range of issues.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

DanielYankura committed:
fixes issue where external variables lagged by one timestep

-issue was caused by external variables never being initialized
-in InitProblemAction.C it now checks if there is an external variable
-if there is then it calls initializeVariable
-initializeVariable was added to BasicExternalProblem and SyncTestExternalProblem

closes #18396

DanielYankura committed:
added back in newline what was deleted previously

GiudGiud commented on test/src/problems/BasicExternalProblem.C:
```diff
@@ -33,6 +33,23 @@ BasicExternalProblem::addExternalVariables()
+  for (const auto elem_ptr : mesh.element_ptr_range())
+  {
+    auto dof_idx = elem_ptr->dof_number(sys_number, _heat_source_var, 0);
+    solution.set(dof_idx, 12345);
```
?

GiudGiud replied:
I would use the ExternalProblem directly rather than those test objects to test the fix

aprilnovak replied:
I'm not sure this is the fix I had in mind -- we don't necessarily have an initial value to put into this external variable until we run our OpenMC simulation. The problem is that even after we run OpenMC, the user objects are lagged by one time step (with the current MOOSE, not including any changes in this PR). This fix seems to be just manually adding an insertion point to initialize the variable (which for us would mean an extra OpenMC run -- as opposed to MOOSE just realizing we already ran OpenMC and properly keeping things up to date).

Is my understanding correct?

DanielYankura replied:
I believe you're correct. When I looked at the two input files you included it looked like for the external case heat_source was never initialized and so it defaulted to 0. Then in the first iteration layered_integral was calculated before heat_source was ever assigned a value which is why it was off for that first timestep. That's why I assumed it was an issue with initialization.

DanielYankura replied:
So I did some more digging and I think I may have changed the execute_on parameter during some of my debugging but forgot to change it back. So I went back and ran the two original input files you included and when both the aux kernel and user object are set to execute on timestep_end I'm getting the same results for both. The only difference being that for the external user object heat_source is 0 for t = 0 unless I initialize heat_source. I'm wondering if this issue may have been resolved in a different PR and adding an option to initialize and external variable was the only thing left to fix? I was going to try and do some testing with a time dependent problem and see if it still works.

DanielYankura replied:
@aprilnovak So I ran similar problems that were time-dependent and both the external and non-external user object versions gave the same results. The only difference was that for the external user object, heat_source is initialized to 0. I think this issue may have been resolved earlier. Unfortunately I'm not sure which PR may have fixed it. For reference here are the two examples I ran. For the external one I modified the SyncSolutions function in BasicExternalProblem.C so that it set heat_source to 12345*time(). If you run into any similar issues let me know and I'll take another look!
[time-depenent example.zip](https://github.com/user-attachments/files/18170973/time-depenent.example.zip)

GiudGiud commented on framework/src/actions/InitProblemAction.C:
```diff
@@ -28,4 +28,9 @@ InitProblemAction::act()
     _problem->init();
   else
     mooseError("Problem doesn't exist in InitProblemAction!");
+
+  // check for external variable to initialize
+  auto external_problem_ptr = std::dynamic_pointer_cast<ExternalProblem>(_problem);
+  if (external_problem_ptr)
+    external_problem_ptr->initializeVariable();
```
any chance you can move that call to initializeVariable inside the ExternalProblem init() ?

GiudGiud replied:
then you wont need to define these new functions to the two test objects

DanielYankura replied:
So the issue I would run into there is that the FEProblemBase object _problem has not been initialized yet. Is there another way I can get a pointer to the external problem object at that point?

GiudGiud replied:
I think it's inited you can see it right above
```
    _problem->init();
```

DanielYankura replied:
Yes, but you wanted it moved into ExternalProblem.C right?

GiudGiud replied:
_problem here is your local problem, of type ExternalProblem

DanielYankura replied:
I see what you're saying. If I call initializeVariable inside of the ExternalProblem constructor, then I won't be able to override it in any derived classes though right?

DanielYankura replied:
Or I can probably put all the logic for initializing variables in the ExternalVariable class and just not try to override it.

DanielYankura replied:
I guess I should clarify, it can be overridden in a derived class but the parent constructor is not going to call the overridden version.

GiudGiud replied:
the _problem->init call is not the constructor
you just need to redefine ExternalProblem::init to include the logic you want to add here

lindsayad replied:
You can call down to the base `FEProblemBase::init` from within your derived `ExternalProblem::init` `override`

lindsayad replied:
but sounds like we're not going to end up going down this path anyways

DanielYankura replied:
That's what I ended up doing. I was going to push it this morning but as you said I'm not sure that this is the right solution.

GiudGiud commented on framework/include/problems/ExternalProblem.h:
```diff
@@ -46,4 +46,9 @@ class ExternalProblem : public FEProblemBase
+  /**
+   * Method to initialize variable before starting the simulation
+   */
+  virtual void initializeVariable() {}
```
```suggestion
  virtual void initializeVariables() {}
```

GiudGiud commented on framework/include/problems/ExternalProblem.h:
```diff
@@ -46,4 +46,9 @@ class ExternalProblem : public FEProblemBase
   virtual void addExternalVariables() {}
+
+  /**
+   * Method to initialize variable before starting the simulation
```
```suggestion
   * Method to initialize variables before starting the simulation
```

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2594319/), step Docs: sync website on b05630b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29520/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2594334/), step Generate coverage on b05630b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c900ee</th>
      <th colspan=4>#29520 b05630</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.17%</td>
      <td>85.16%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108183</td>
      <td>108176</td>
      <td>-7</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18844</td>
      <td>18855</td>
      <td>+11</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29520/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29520/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29520/coverage/xfem)

</details>



This comment will be updated on new commits.

DanielYankura committed:
Various changes for PR

-Reverted BasicExternalProblem and SynceTestExternalProblem
-Check to external variable initialization moved to FEProblemBase.C
-function name changed to initializeVariables
-initializeVariables can be modified in ExternalProblem.C but can also be overriden if needed

DanielYankura committed:
style changes

DanielYankura committed:
Removed time dependence from BasicExternalProblem that I used for a previous test
